lualib/resty/events/worker.lua

```diff
@@ -243,21 +244,23 @@ function _M:communicate(premature)
             sleep(0)
 
             ::continue::
-        end -- while not exiting
-    end)  -- events_thread
+        end -- while not exiting or restarting
+    end) -- events_thread
 
     local ok, err, perr = wait(write_thread, read_thread, events_thread)
 
-    kill(write_thread)
     kill(read_thread)
-    kill(events_thread)
-
-    self._connected = nil
+    kill(write_thread)
 
     if exiting() then
+        kill(events_thread)
         return
     end
 
+    -- killing events thread is not safe in case we are restarting the timer
+    stop_handling_events = true
+    wait(events_thread)
```

Comment:
Should we also check the response of `wait` in case the handler returns any error.